apps/studio/src/services/plugin/PluginFileManager.ts

```diff
@@ -0,0 +1,241 @@
+  let report: DownloaderReport;
+
+  try {
+    const downloader = new Downloader(options);
+    options.signal?.addEventListener("abort", downloader.cancel);
+    report = await downloader.download().finally(() => {
+      options.signal?.removeEventListener("abort", downloader.cancel);
+    });
+    downloadStatus = report.downloadStatus;
+  } catch (e) {
+    if (e.responseBody) {
+      throw new PluginDownloadError(e.responseBody, e);
+    }
+    throw new PluginDownloadError("Download failed", e);
+  }
+
+  if (downloadStatus === "ABORTED") {
+    const error = new PluginDownloadError("Download aborted");
+    error.status = "ABORTED";
+    throw error;
+  }
+
+  return report.filePath;
+}
+
+export default class PluginFileManager {
+  constructor(private readonly repositoryService: PluginRepositoryService) {}
+
+  /** Download all plugin files to `directory` */
+  async download(
+    entry: PluginRegistryEntry,
+    release: Release,
+    options: {
+      signal?: AbortSignal;
+      /** for update */
+      tmp?: boolean;
+    } = {}
+  ) {
+    const directory = this.getDirectoryOf(entry.id);
+
+    try {
+      fs.mkdirSync(directory, { recursive: true });
+
+      log.debug(
+        `Downloading plugin "${entry.id}" version "${release.version}"...`
+      );
+
+      log.debug(
+        `Downloading plugin manifest: "${release.manifestDownloadUrl}"...`
+      );
+      await download({
+        url: release.manifestDownloadUrl,
+        directory,
+        signal: options.signal,
+        fileName: options.tmp
+          ? PLUGIN_MANIFEST_TMP_FILENAME
+          : PLUGIN_MANIFEST_FILENAME,
+      });
+
+      log.debug(`Downloading plugin script: "${release.scriptDownloadUrl}"...`);
+      await download({
+        url: release.scriptDownloadUrl,
+        directory,
+        signal: options.signal,
+        fileName: options.tmp
+          ? PLUGIN_SCRIPT_TMP_FILENAME
+          : PLUGIN_SCRIPT_FILENAME,
+      });
+
+      log.debug(`Downloading plugin style: "${release.styleDownloadUrl}"...`);
+      await download({
+        url: release.styleDownloadUrl,
+        directory,
+        signal: options.signal,
+        fileName: options.tmp
+          ? PLUGIN_STYLE_TMP_FILENAME
+          : PLUGIN_STYLE_FILENAME,
+      }).catch((e) => {
+        if (e instanceof PluginDownloadError && e.status === "NOT_FOUND") {
+          log.debug("Plugin style not found. Skipping.");
+          return;
+        }
+        throw e;
+      });
+    } catch (e) {
+      fs.rmSync(directory, { recursive: true, force: true });
+      log.debug("Download failed", e);
+      throw e;
+    }
+  }
+
+  async update(
+    entry: PluginRegistryEntry,
+    release: Release,
+    options: { signal?: AbortSignal } = {}
+  ) {
+    await this.download(entry, release, { ...options, tmp: true });
+
+    const directory = this.getDirectoryOf(entry.id);
+
+    // Rename old files
+    fs.rmSync(path.join(directory, PLUGIN_MANIFEST_FILENAME), { force: true });
```

Comment:
Don't use the `sync` versions of these -- it'll lock up the process, best to use the promises API, it's already an async function